client/app/components/SearchBar/SearchBar.js

```diff
@@ -13,16 +15,139 @@ const SEARCH_MODES = [
+    if (status !== serviceStatus.OK) {
+      reject(new Error(`Could not get details for place id "${placeId}"`));
+    } else {
+      resolve(place);
+    }
+  });
+});
+
+const getPrediction = (location) => new Promise((resolve, reject) => {
+  const serviceStatus = window.google.maps.places.PlacesServiceStatus;
+  const service = new window.google.maps.places.AutocompleteService();
+
+  service.getPlacePredictions({ input: location }, (predictions, status) => {
+    if (status !== serviceStatus.OK) {
+      reject(new Error(`Prediction service status not OK: ${status}`));
+    } else if (predictions.length === 0) {
+      reject(new Error(`No predictions found for location "${location}"`));
+    } else {
+      resolve(getDetails(predictions[0].place_id));
+    }
+  });
+});
+
 class SearchBar extends Component {
   constructor(props) {
     super(props);
 
-    // The values of the inputs are synced to the component state
-    // until the form is submitted.
-    this.state = {
-      keyword: '',
-      location: '',
-    };
+    // Bind `this` within methods
+    this.handleSubmit = this.handleSubmit.bind(this);
+
+    this.state = { selectedPlace: null };
+  }
+  componentDidMount() {
+    document.body.classList.add(css.topLevelBody);
+    if (!this.locationInput) {
+      return;
+    }
+    const bounds = { north: -90, east: -180, south: 90, west: 180 };
+    const autocomplete = new window.google.maps.places.Autocomplete(this.locationInput, { bounds });
+    autocomplete.setTypes(['geocode']);
+    this.placeChangedListener = window.google.maps.event.addListener(
+      autocomplete,
+      'place_changed',
+      () => this.setState({ selectedPlace: autocomplete.getPlace() })
+    );
+  }
+  componentWillUnmount() {
+    document.body.classList.remove(css.topLevelBody);
+
+    if (this.placeChangedListener) {
+      this.placeChangedListener.remove();
+    }
+
+    // Clean up the generated autocompletion UI. Assumes there is only one.
+    const container = document.body.querySelector('.pac-container');
+
+    if (container) {
+      document.body.removeChild(container);
+    }
+  }
+  handleSubmit() {
+    if (!this.keywordInput && !this.locationInput) {
+      throw new Error('No input refs saved to submit SearchBar form');
+    }
+
+    const keywordValueStr = this.keywordInput ? this.keywordInput.value.trim() : '';
+    const locationValueStr = this.locationInput ? this.locationInput.value.trim() : '';
+
+    if ((keywordValueStr + locationValueStr).length === 0) {
+      // Skip submit when all inputs are empty
+      return;
+    }
+
+    const onSubmit = this.props.onSubmit;
+
+    if (!this.locationInput) {
+      // Only keyword input, submitting current input value
+      onSubmit({
+        keyword: keywordValueStr,
+        location: null,
+        coordinates: null,
+      });
+      return;
+    }
+
+    const keywordValue = this.keywordInput ? this.keywordInput.value : null;
```

Comment:
Notice the clear distinction of the empty String vs. `null` value in the `onSubmit` callback. Empty value means that the input is in use but the value was empty, `null` means that the input is not in use. This way the returned object shape is always the same, but the caller can still differentiate which inputs provided some actual values.